ci: composable avalanchego action

Elvis339 opened the pull request:
## Why this should be merged
Enables composable CI by allowing repositories to setup AvalancheGo with custom dependency versions. This solves the problem where teams need AvalancheGo context for testing but can't easily compose different repo versions in their CI workflows.

[More context](https://docs.google.com/document/d/1ynQmmt7U_77PzojlckvZcMfBG3wMXA4o64zDGckLwxs/edit?usp=sharing)

https://github.com/ava-labs/avalanchego/blob/6de7ef24da194951a670ad8f3c96999471a15e7f/scripts/setup_firewood.sh#L34

## How this works
Prepares build environment with dependency replacements for custom workflows

## How this was tested
**Test Run**: https://github.com/ava-labs/avalanchego/actions/runs/18103163342/job/51511123516

**Workflow**: https://github.com/ava-labs/avalanchego/blob/1674dfaf1135cc21d63c4a8a6bfbaee6c915b36a/.github/workflows/ci.yml#L285

This test validates the setup action by configuring Firewood v0.0.12 and v0.0.13, then running the c-chain-reexecution benchmark. The setup phase works correctly - the benchmark failure is unrelated and outside this PR's scope.

**Note**: The [setup script](https://github.com/ava-labs/avalanchego/blob/6de7ef24da194951a670ad8f3c96999471a15e7f/scripts/setup_firewood.sh#L34) depends on a custom branch in [Firewood](https://github.com/ava-labs/firewood/blob/composable-ci-action/.github/scripts/build.sh).

## Need to be documented in RELEASES.md?
No. Internal CI change

Elvis339 committed:
ci(avalanchego-build)

Elvis339 committed:
ci: temp. run custom action on github runners

Elvis339 committed:
ci: temp. run custom action on github runners with ubuntu-latest tag

Elvis339 committed:
ci(gh_runners): temp. checkout

Elvis339 committed:
ci: remove checkout

Elvis339 committed:
ci: add checkout back

Elvis339 committed:
ci: remove defaults from coreth and libevm

Elvis339 committed:
debug

Elvis339 committed:
ci: use $GITHUB_ACTION_PATH

Elvis339 committed:
ci: github runners

Elvis339 committed:
ci: github runners use nix and use realpath

Elvis339 committed:
ci: github runners run help from bin

Elvis339 committed:
ci: self-hosted runners

Elvis339 committed:
ci: self-hosted runners install build-essential

Elvis339 committed:
ci: self-hosted runners run & build reexecution bench

Elvis339 committed:
ci: self-hosted runners run & build reexecution bench

Elvis339 committed:
ci(reexec): temp. disable go setup

Elvis339 committed:
re-run re-exec

Elvis339 committed:
ci: set permissions

Elvis339 committed:
ci(avalanchego-build-action): use composable Firewood action

Elvis339 committed:
ci: reexecution bench with firewood custom action

Elvis339 committed:
ci: add back ci jobs & remove tmp. jobs

Elvis339 committed:
Merge branch 'master' into composable-ci-action

Elvis339 committed:
ci: uncomment c-chain-reexecution-benchmark-* pull_request directive

Elvis339 committed:
chore: address copilot PR review

Elvis339 commented on .github/actions/c-chain-reexecution-benchmark/action.yml:
```diff
@@ -98,6 +98,7 @@ runs:
         prometheus_password: ${{ inputs.prometheus-password }}
         grafana_dashboard_id: 'Gl1I20mnk/c-chain'
         runtime: "" # Set runtime input to empty string to disable log collection
+        prometheus_url: ${{ inputs.prometheus-push-url }}
```
It's one of the required parameters in `run-monitored-tmpnet-cmd` and it used to control starting of metrics collection that has been fixed so this is no longer needed I just need to merge master once that patch gets in.

Elvis339 committed:
ci(c-chain-reexecution-benchmark)

aaronbuchwald commented on .github/actions/avalanchego-build-action/action.yml:
```diff
@@ -0,0 +1,120 @@
+    - name: Setup Firewood FFI
+      if: inputs.firewood != ''
+      id: firewood
+      uses: ava-labs/firewood/.github/actions/build-action@composable-ci-action
```
Where is this action defined? I don't see it on Firewood main

Elvis339 commented on .github/actions/avalanchego-build-action/action.yml:
```diff
@@ -0,0 +1,120 @@
+          go mod edit -replace github.com/ava-labs/libevm=../libevm
+        fi
+
+        go mod tidy
+        go mod download
```
Yes it will, it should probably be cleaned up. Will handle that case.

aaronbuchwald replied:
I think we could probably do less here and only support setup rather than adding a script that builds different binaries once setup has completed.

aaronbuchwald commented on .github/actions/avalanchego-build-action/README.md:
```diff
@@ -0,0 +1,101 @@
+    coreth: "my-feature-branch"
+    libevm: "experimental-branch"
+
+- name: Run custom build commands
+  run: |
+    cd build/avalanchego
+    ./scripts/run_task.sh reexecute-cchain-range
+    ./scripts/run_task.sh my-custom-task
```
This would not work because the current task definition uses `go run`, so it would not use the the compiled test binary.

Elvis339 replied:
It works here.

- [Workflow](https://github.com/ava-labs/avalanchego/blob/dfa8c8164e7578d2bdec93a5c1595dffcdf1732e/.github/workflows/ci.yml#L285)
- [Run](https://github.com/ava-labs/avalanchego/actions/runs/18019895212/job/51274339039)

Elvis339 committed:
ci(avalanchego-build-action): use `go get` instead of checking libevm and coreth versions locally

maru-ava reviewed (commented):
Why was this not implemented as a task that could be executed both locally and in CI? A task has the advantage of being easier to implement and maintain - since iteration doesn't require cycling through CI - and then it would be possible to trivially reproduce a binary without CI.

Elvis339 commented:
> Why was this not implemented as a task that could be executed both locally and in CI?

The Action approach was chosen for cross-repository consumption. But now that you mention Task I guess it could be made to to run locally and in CI and be exposed as GH action for cross-repository consumption.

JuanLeon2 commented on .github/actions/avalanchego-build-action/action.yml:
```diff
@@ -0,0 +1,120 @@
+    required: false
+    default: ''
+
+outputs:
```
Can we also output the artifact name (rather than solely relying on the convention described in README.md)

Elvis339 committed:
feat(build): extend build with Firewood

Elvis339 committed:
ci(avalanche-go-setup):
- remove build mode in favor of composability
- use scripts allowing both local and ci workflow

Elvis339 committed:
ci: test composition

Elvis339 committed:
ci: test composition - switch AWS role

Elvis339 committed:
ci: test composition - switch AWS role

Elvis339 committed:
ci: test composition - set permissions

Elvis339 committed:
ci: test composition - set permissions

Elvis339 committed:
ci: test composition, set `config: firewood` to reexec bench

Elvis339 committed:
ci: set current state dir src to firewood 100

Elvis339 committed:
ci: revert to default

Elvis339 commented:
Switched to task that can execute both locally and in CI.

- Extended `build.sh` with `-f` option signaling Firewood version
- Added new [script](https://github.com/ava-labs/avalanchego/blob/6de7ef24da194951a670ad8f3c96999471a15e7f/scripts/setup_firewood.sh) `setup_firewood.sh` 
- Refactored CI to only support setting Avalanchego without building targets

Elvis339 committed:
Merge branch 'master' into composable-ci-action

maru-ava reviewed (commented):
Given that coreth is the integration point for firewood, and coreth is only compatible with specific versions of avalanchego, I'm not sure this open-ended approach (supporting independent configuration of the versions of firewood, coreth and avalanchego) makes sense.

I'm also not sure why avalanchego should be the integration point. Coreth already runs avalanchego e2e tests for its pinned version of avalanchego. So maybe that same mechanism can be extended to support a coreth job for each of the targeted versions of firewood?

Elvis339 commented on .github/actions/avalanchego-setup-action/action.yml:
```diff
@@ -0,0 +1,53 @@
+runs:
+  using: 'composite'
+  steps:
+    - name: Checkout AvalancheGo
```
That's a fair point. I can see the appeal of consolidating all go.mod updates in one place.

I'm thinking of keeping coreth/libevm updates explicit in the action makes the dependency changes visible and optional. They're simple `go get` commands, so putting  them in setup_firewood.sh would add context/coupling without clear benefit - the script would need to handle optional arguments for dependencies it doesn't own.

That said, I'm not strongly opposed to consolidating them. Do you see a clear advantage to moving them into the script? Or is this more about having a consistent pattern for where dependency management happens?

Happy to restructure if there's a compelling reason I'm missing.

Elvis339 committed:
ci(avalanchego-setup-action): add `firewood-path` checkout path allowing Firewood runtime to run on nvme

Elvis339 committed:
Merge remote-tracking branch 'origin/composable-ci-action' into composable-ci-action

Elvis339 committed:
ci(avalanchego-setup-action): handle checkout-path

Elvis339 committed:
ci(avalanchego-setup-action): handle checkout-path

Elvis339 committed:
ci(avalanchego-setup-action): correct Firewood setup path

Elvis339 committed:
Merge branch 'master' of github.com:ava-labs/avalanchego into composable-ci-action

Elvis339 committed:
chore: externalize c-chain-reexecution json based workflow into tasks

Elvis339 committed:
ci(c-chain-reexecution): remove relative imports to make this action reusable

Elvis339 committed:
ci(c-chain-reexecution): use correct shells

Elvis339 committed:
ci(c-chain-reexecution)!:

Elvis339 committed:
copilot PR comments

Elvis339 committed:
ci(c-chain-reexecution): set absolute path to benchmark output file resolving error on compare benchmark results job

Elvis339 committed:
address pr review

Elvis339 committed:
address PR

Elvis339 committed:
ci(c-chain-reexecution-benchmark): revert change on `Push Post-State to S3`

Elvis339 committed:
Merge branch 'master' into es/task-reexecution

Elvis339 committed:
Merge branch 'master' into es/task-reexecution

Elvis339 committed:
Merge branch 'master' into es/task-reexecution

Elvis339 committed:
chore(benchamrk_cchain_range): avoid duplicating the default of `RUNNER_NAME`

Elvis339 committed:
ci(c-chain-reexecution):
- inline set task env
- remove unrelated `TMPNET_START_METRICS_COLLECTOR` env var

Elvis339 committed:
Merge branch 'master' into es/task-reexecution

Elvis339 committed:
chore: rm doc

Elvis339 committed:
Merge remote-tracking branch 'origin/es/task-reexecution' into es/task-reexecution

Elvis339 committed:
Merge branch 'master' into es/task-reexecution

Elvis339 committed:
ci(c-chain-reexecution): support alternative approach to task based operations

Elvis339 committed:
ci(c-chain-reexecution): support both container and native altenrative (custom) approach to task based operations

Elvis339 committed:
Merge remote-tracking branch 'origin/es/task-reexecution' into es/task-reexecution

Elvis339 committed:
lint

Elvis339 committed:
refactor(scripts): extract S3 bucket config from copy_dir.sh to Taskfile env making copy_dir.sh reusable for any S3 bucket

Elvis339 committed:
fix

Elvis339 committed:
Merge branch 'master' into es/task-reexecution

Elvis339 committed:
ci(c-chain-reexecution): use dynamic env variable for benchmark output file path

Elvis339 committed:
Merge remote-tracking branch 'origin/es/task-reexecution' into es/task-reexecution

Elvis339 committed:
ci(c-chain-reexecution): remove unused `TASK_NAME` environment variable

Elvis339 committed:
ci(c-chain-reexecution): enforce required `runner` parameter by removing default options

Elvis339 committed:
Merge branch 'master' into es/task-reexecution

Elvis339 committed:
Merge branch 'master' into composable-ci-action

Elvis339 committed:
Merge branch 'es/task-reexecution' into composable-ci-action

Elvis339 committed:
chore: setup firewood

Elvis339 committed:
chore(setup-firewood): use correct branch for cloning

Elvis339 committed:
ci: disable ci jobs

Elvis339 committed:
run on pull request

Elvis339 committed:
t

Elvis339 committed:
stash changes

Elvis339 committed:
fix

Elvis339 committed:
ci: support only self-hosted

Elvis339 committed:
ci: setup go

joshua-kim commented:
Closing. If this was closed in error please re-open.
